### ISSUE TITLE
groupby_dynamic causes memory allocation crash with mixed index and interval types

### ISSUE THREAD

**Issue by ptomecek** [pola-rs/polars#2438]
#### Are you using Python or Rust?

Python

#### Which feature gates did you use?

This can be ignored by Python users.

#### What version of polars are you using?

0.12.13

#### What operating system are you using polars on?

Linux CentOS7 and Windows 10

#### Describe your bug.

I am trying to do a kdb-style window join (https://code.kx.com/q/ref/wj/) in polars, by leveraging groupby_dynamic, where for every element in my (possibly irregular) frame ("1i"), I aggregate over a fixed time window around that observation ("2h"). This causes a memory allocation crash.

#### What are the steps to reproduce the behavior?

```
import polars as pl
from datetime import datetime, timedelta
times = pl.date_range(low=datetime(2021, 12, 16), high=datetime(2021, 12, 16, 4), interval=timedelta(minutes=30))
df = pl.DataFrame({'time': times, 'n': range(len(times))})
out = df.groupby_dynamic('time', every='1i', period='2h').agg([pl.col('time').min(), pl.col('time').max()])
```

#### What is the actual behavior?

```
memory allocation of 691200000000032 bytes failed
```

#### What is the expected behavior?

Either raise NotImplementedError, or ideally, compute the aggregation correctly (min and max time occurring in a two hour window around each original row).

**Comment by ritchie46** [comment]
We should raiae an error you cannot combine "2i" (integers) with times.

Now we make a window every 2 nanoseconds. Exploding menory.

**Comment by ptomecek** [comment]
Thanks for the quick fix! As for requesting this feature, should I put this in a new issue? For example, it's something very straightforward in pandas, but that does not seem to have a polars equivalent

```
import pandas as pd
from datetime import datetime, timedelta
times = pd.date_range(datetime(2021, 12, 16), datetime(2021, 12, 16, 4), freq=timedelta(minutes=30))
df = pd.DataFrame({'n': range(len(times))}, index=times)
df.rolling('2H').agg({'n': ['min', 'max']})
```

**Comment by ritchie46** [comment]
You can! It is being released as we speak in #2441 

Here is an example in polars based on your earlier example:

```python
times = pl.date_range(datetime(2021, 12, 16), datetime(2021, 12, 16, 4), interval=timedelta(minutes=30), )
df = pl.DataFrame({
    "n": range(len(times)),
    "times": times
})

(df.groupby_rolling(index_column="times", period="2h").agg([
    pl.col("n").min(),
    pl.col("n").max(),
    pl.col("n").count()
]))
```

```
shape: (9, 4)
┌─────────────────────┬───────┬───────┬─────────┐
│ times               ┆ n_min ┆ n_max ┆ n_count │
│ ---                 ┆ ---   ┆ ---   ┆ ---     │
│ datetime[ns]        ┆ i64   ┆ i64   ┆ u32     │
╞═════════════════════╪═══════╪═══════╪═════════╡
│ 2021-12-16 00:00:00 ┆ 0     ┆ 0     ┆ 1       │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 2021-12-16 00:30:00 ┆ 0     ┆ 1     ┆ 2       │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 2021-12-16 01:00:00 ┆ 0     ┆ 2     ┆ 3       │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 2021-12-16 01:30:00 ┆ 0     ┆ 3     ┆ 4       │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ ...                 ┆ ...   ┆ ...   ┆ ...     │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 2021-12-16 02:30:00 ┆ 2     ┆ 5     ┆ 4       │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 2021-12-16 03:00:00 ┆ 3     ┆ 6     ┆ 4       │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 2021-12-16 03:30:00 ┆ 4     ┆ 7     ┆ 4       │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ 2021-12-16 04:00:00 ┆ 5     ┆ 8     ┆ 4       │
└─────────────────────┴───────┴───────┴─────────┘
```

Make sure to also check out `groupby_dynamic` which can give you even more control over time windows, offsets, periods and subgroups.

**Comment by ptomecek** [comment]
That's awesome, thanks!
